Feature/010-COMMENT

seokhyeong9171 opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
**AS-IS**
- post의 view, like 정보가 MariaDB, Redis에 중복 저장되었습니다.

**TO-BE**
- post의 view, like 정보를 Redis에만 저장하도록 변경하였습니다
   -  post별 유저의 view, like 정보를 담았던 기존에서 유저 별 post의 view, like 정보를 담도록 수정했습니다.
- post에 comment 기능을 추가하였습니다
   - child comment를 등록할 수 있도록 하였습니다.
   - comment를 삭제할 경우 child comment가 있을 경우 해당 comment의 delete flag만 변경해 주고 child comment가 없을 경우 삭제하였습니다.
- post를 삭제할 경우 관련 comment, view, like 값도 삭제되도록 하였습니다

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [ ] 테스트 코드
- [ ] API 테스트

seokhyeong9171 committed:
refactor: ForumController 이름 변경

seokhyeong9171 committed:
feature: post 조회, 좋아요 기능 컨트롤러 개발

seokhyeong9171 committed:
feature: postApplication

seokhyeong9171 committed:
feature: PostService

seokhyeong9171 committed:
feature: CommentController

seokhyeong9171 committed:
feature: comment 기능

seokhyeong9171 committed:
feature: redis key 수정

seokhyeong9171 committed:
feature: docker 파일 추가

seokhyeong9171 committed:
refactor: 리팩토링

seokhyeong9171 committed:
refactor: ERD 이미지 변경

seokhyeong9171 committed:
feature: post 삭제 시 comment 같이 삭제

seokhyeong9171 committed:
refactor: 리팩토링

seokhyeong9171 committed:
refactor: 리팩토링

seokhyeong9171 committed:
refactor: 리팩토링

seokhyeong9171 committed:
refactor: 리팩토링

mingming-mentor commented on domain/src/main/java/com/health/domain/entity/CommentEntity.java:
```diff
@@ -34,4 +42,48 @@ public class CommentEntity extends BaseEntity{
     private UserEntity createdUser;
+
+    @Column(name = "delete_yn")
+    private Boolean isDeleted;
```
deleted라는 네이밍도 좋을 것 같아요. 그러면 추후에 isDeleted()라는 메소드를 이용해서 getter를 쓰는게 좀더 자연스럽거든요.

mingming-mentor commented on forum-service/src/main/java/com/health/forumservice/service/impl/CommentServiceImpl.java:
```diff
@@ -0,0 +1,201 @@
+
+    // 오래된 순서대로 정렬
+    return getPostByPostId(postId).getCommentList().stream()
+        .filter(c -> !c.getIsDeleted())
```
삭제된 comment를 select시에 가지고올일이 아예 없다면 @Where를 써도 괜찮을 것 같아요
https://jaime-note.tistory.com/450

mingming-mentor commented on forum-service/src/main/java/com/health/forumservice/service/impl/CommentServiceImpl.java:
```diff
@@ -0,0 +1,201 @@
+    validateDeleted(findComment);
+    validateRightPost(findPost, findComment);
+
+    // child comment 없는 경우엔 comment 삭제
+    // child comment 있는 경우엔 delete flag만 변경
```
child comment 존재 여부에 따라 다르게 동작하는 이유가 있을까요?

seokhyeong9171 replied:
@mingming-mentor child comment같은 경우에는 parent comment와 작성자가 다른 경우이어서 parent comment를 삭제했다 하더라도 child comment도 같이 삭제되는건 아닌것 같다고 판단했습니다!

mingming-mentor replied:
@seokhyeong9171 님, 그러면 parent comment가 삭제는 되어야하지만 하위의 child comment가 있으므로 FK 관계 때문에 parent comment를 남겨둔 상태로 이해하면 될까요? 

어떤 경우는 row가 삭제되고 어떤 경우는 deleted flag 처리가 되는데, 사실 그냥 딱 설계만 보았을땐 deleted flag가 있기 때문에 실제 row가 삭제된다고는 생각이 안들것 같아서요. 

추가적으로, 만약 parent comment가 삭제가 되고 child comment는 있는 경우, post에 들어갔을때 child comment는 보이게 되는건가요? 
그런거라면 제 생각에는 hard delete를 두지 않고 soft delete 방식으로만 처리를 하는 방식도 좋아보여요. 
혹은 FK를 지우고, parent comment가 삭제가 되면 삭제를 시키고, 대신 comment에 status를 두어서 child comemnt의 경우 PARENT_DELETED 같은 상태를 두어서 parent comment 가 삭제된 경우임을 표기해주어도 될것 같구요.

요지는, deleted flag가 어떤 경우 표기가 되는가가 명확하지 않아보여서 추후 협업시에 실수가 일어날 수 있는 구조인것 같아요. 이 부분이 조금 개선되면 어떨까 의견드려봅니다

seokhyeong9171 replied:
@mingming-mentor 
원래는 parent comment를 삭제하면 child comment는 보이게 하기 위해서 기존 방식을 적용한거 였는데요 
말씀 해주신거 듣고 생각해보니깐 조금 부족한 부분이 있었던것 같네요
comment들은 flag를 두어서 모두 soft delete 하는 방식으로 처리하겠습니다!

mingming-mentor commented on forum-service/src/main/java/com/health/forumservice/service/impl/CommentServiceImpl.java:
```diff
@@ -0,0 +1,201 @@
+
+    findChildComment.updateComment(domainForm);
+
+    return null;
```
update시에 null을 return 하는 이유가 있을까요? 굳이 return 값이 필요없는 경우라면 void로 return type을 바꿔줘도 좋을 것 같아요

mingming-mentor reviewed (approved):
뭔가 제 피드백이 잘못 전달된것 같은데요, post_like table은 존재해야하지 않을까요..? 
Redis의 경우 메모리 캐시이기 때문에 데이터가 날라갈 가능성이 있습니다. 따라서 항상 persistent database와 병행해서 사용해야합니다. 

제가 드린 피드백은, 
기존에는 post_like 데이터가 원본 데이터라고 가정했을때, 이를 post 별로 계산하여 저장한 like column이 post table에 존재했었는데요. 이 데이터가 존재해야할 이유는 정렬시에 like 값을 이용한다던가 등에서 사용되기 때문이라고 생각되어요. 그렇지 않다면 post_like에서 연산하여 사용해도 되니깐요. 
그런데 말씀하시기로는 이런 정렬을 redis를 사용하신다고 하셨으니 post에서 해당 컬럼을 삭제하는 것이 어떤가 말씀드린것이었습니다.
같은 데이터가 3개 저장되어있지만, 이중 원본인 post_like는 반드시 존재해야하고, 이를 연산을 위해 저장하는 redis or post table의 like 컬럼중에서는 하나만 존재하면 될것 같아요

seokhyeong9171 commented:
@mingming-mentor 아 제가 피드백을 잘못 이해 했었나보네요... 다시 수정해 보겠습니다 감사합니다!

seokhyeong9171 committed:
refactor: 피드백 반영

comment 삭제 시 soft delete로 변경,
delete flag 필드명 변경,
post view, like 값 persistent database에도 저장

seokhyeong9171 commented:
@mingming-mentor  멘토님 피드백 주신것들 참고해서 반영해 보았는데 한번만 더 피드백 부탁드립니다

seokhyeong9171 committed:
refactor: 리팩토링

mingming-mentor reviewed (approved):
수고 많으셨습니다!!
